### PR TITLE
rework and align refcount management

### PR DESCRIPTION
**Pull request by mandelsoft**
**What this PR does / why we need it**:

Provided elements in contexts OCM and OCI now support a uniform reference management
to release allocated resources after the close of the last using viw object, like
Repositories, ComponetAccesses, NamespaceAccesses, ArtifactAccesses and ComponentVersioAccesses.

A nw `resource` package now provided a generic handling of consumer-facing view objects, underlying implementation objects and reference managers.

The support implementation if OCI and OCm contexts has been modified to use use this framework to achieve a uniform reference handling.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release Notes**:

```feature user

```

### CODE REVIEW

**Review by fabianburth — CHANGES_REQUESTED** [review]

File: pkg/common/accessio/resource/resource.go [hunk 1]
```diff
@@ -0,0 +1,311 @@
+// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Open Component Model contributors.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+// Package resource provided some support to implement
```
[the remark this answers is not in the document]
Comment (mandelsoft) [reply pkg/common/accessio/resource/resource.go:5]: done

File: pkg/common/accessio/resource/resource.go [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+// closeable backing resources featuring multiple
+// separately closeable references. The backing resource
+// is finally closed, when the last reference is closed.
+// hereby, the reference implements the intended resource
```
[the remark this answers is not in the document]
Comment (mandelsoft) [reply pkg/common/accessio/resource/resource.go:9]: done

File: pkg/common/accessio/resource/resource.go [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+// interface including the reference related part, which
+// includes a Dup method, which can be used to gain a
+// new additional reference to the backing object.
+//
+// Those references are called View in the package.
```
[the remark this answers is not in the document]
Comment (mandelsoft) [reply pkg/common/accessio/resource/resource.go:14]: done

File: pkg/common/accessio/resource/resource.go [hunk 4]
```diff
@@ -0,0 +1,311 @@
+// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Open Component Model contributors.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+// Package resource provided some support to implement
+// closeable backing resources featuring multiple
```
[the remark this answers is not in the document]
Comment (mandelsoft) [reply pkg/common/accessio/resource/resource.go:6]: backing resource here means the common object behind the views carrying the ref.  This one then may keep access to external resources, which may be released if the last ref is vanishing. (the backing object is closed). I adjusted the text accordingly.

File: pkg/common/accessio/resource/resource.go [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+// separately closeable references. The backing resource
```
[the remark this answers is not in the document]
Comment (mandelsoft) [reply pkg/common/accessio/resource/resource.go:7]: see above

File: pkg/common/accessio/resource/resource.go [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+	View(main ...bool) (T, error)
+	IsClosed() bool
+}
+
+// ResourceViewCreator is a function which must be provided by the resource provider
+// to map an implementation to the resource interface T.
+// It must use NewView to create the view related part of a resource.
+type ResourceViewCreator[T any, I io.Closer] func(I, CloserView, ViewManager[T]) T
+
+////////////////////////////////////////////////////////////////////////////////
```
[the remark this answers is not in the document]
Comment (mandelsoft) [reply pkg/common/accessio/resource/resource.go:150]: I moved it up.

File: pkg/common/accessio/resource/resource.go [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+	return i.creator(i.impl, v, i), nil
+}
+
+func (i *viewManager[T, I]) IsClosed() bool {
+	return i.refs.IsClosed()
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+type noneRefCloser[T io.Closer] struct {
```
[the remark this answers is not in the document]
Comment (mandelsoft) [reply pkg/common/accessio/resource/resource.go:195]: done

File: pkg/common/accessio/resource/resource.go [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+////////////////////////////////////////////////////////////////////////////////
+
+type resourceView[T any] struct {
+	view CloserView
+	mgr  ViewManager[T]
+}
+
+// NewView is to be called by a resource view creator to map
+// the given resource implementation to complete resource interface.
+// It should, create an object with two local embedded fields:
```
[the remark this answers is not in the document]
Comment (mandelsoft) [reply pkg/common/accessio/resource/resource.go:233]: done

File: pkg/common/accessio/resource/resource.go [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+//   - the returned ResourceView and the
+//   - given resource implementation.
+func NewView[T resourceViewInterface[T]](v CloserView, d ViewManager[T]) ResourceViewInt[T] {
+	return &resourceView[T]{v, d}
+}
+
+// NewNonRefView provides a reference-less view directly for the reference manager.
+// It is valid as long as the reference manger is not closed with the last regular
```
[the remark this answers is not in the document]
Comment (mandelsoft) [reply pkg/common/accessio/resource/resource.go:241]: done

**Review by fabianburth — APPROVED** [review]

**Pull request merged**